src/main/java/com/google/devtools/build/lib/syntax/MethodDescriptor.java

```diff
@@ -59,19 +62,157 @@ private MethodDescriptor(
       boolean allowReturnNones,
       boolean useStarlarkThread,
       boolean useStarlarkSemantics) {
+    boolean isStringModule = method.getDeclaringClass() == StringModule.class;
+
     this.method = method;
     this.annotation = annotation;
     this.name = name;
     this.doc = doc;
     this.documented = documented;
     this.structField = structField;
-    this.parameters = parameters;
+    this.parameters =
+        !isStringModule ? parameters : Arrays.copyOfRange(parameters, 1, parameters.length);
     this.extraPositionals = extraPositionals;
     this.extraKeywords = extraKeywords;
     this.selfCall = selfCall;
     this.allowReturnNones = allowReturnNones;
     this.useStarlarkThread = useStarlarkThread;
     this.useStarlarkSemantics = useStarlarkSemantics;
+    try {
+      MethodHandle mh = MethodHandles.lookup().unreflect(method);
+
+      if (isStringModule) {
+        // String methods get the string as an extra argument
+        // because their true receiver is StringModule.INSTANCE.
+        mh = MethodHandles.insertArguments(mh, 0, StringModule.INSTANCE);
```

Comment:
It might be possible to dispense with INSTANCE entirely (and make all the StringModule methods static). It exists only as a dummy so that every call was an instance method, but I think the MethodHandle machinery doesn't really care about the difference between instance methods and explicit parameters, which is nice.

Reply:
Not sure if it worth the trouble (need to have special handling in annotation validation, and it won't give perf benefits), but possible.